realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -897,24 +897,139 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeNotEqual__J_3JLja
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringNotEqual);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBeginsWith(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeBeginsWith__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringBeginsWith);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEndsWith(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeEndsWith__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringEndsWith);
 }
 
-JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeContains(
+JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeContains__J_3JLjava_lang_String_2Z(
     JNIEnv* env, jobject, jlong nativeQueryPtr, jlongArray columnIndexes, jstring value, jboolean caseSensitive)
 {
     TableQuery_StringPredicate(env, nativeQueryPtr, columnIndexes, value, caseSensitive, StringContains);
 }
 
+// Binary
+
+enum BinaryPredicate {
+    BinaryEqual,
+    BinaryNotEqual,
+    BinaryContains,
+    BinaryBeginsWith,
+    BinaryEndsWith
+};
+
+static void TableQuery_BinaryPredicate(JNIEnv *env, jlong nativeQueryPtr, jlongArray columnIndices, jbyteArray value, BinaryPredicate predicate) {
+    JniLongArray arr(env, columnIndices);
+    jsize arr_len = arr.len();
+    try {
+        BinaryData value2;
+        if (value == NULL) {
+            if (!TBL_AND_COL_NULLABLE(env, getTableByArray(nativeQueryPtr, arr).get(), arr[arr_len-1])) {
+                return;
+            }
+            value2 = BinaryData();
+        }
+        else {
+            jbyte* bytePtr = env->GetByteArrayElements(value, NULL);
+            if (!bytePtr) {
+                ThrowException(env, IllegalArgument, "binaryPredicate");
+                return;
+            }
+            size_t dataLen = S(env->GetArrayLength(value));
+            value2 = BinaryData(reinterpret_cast<char*>(bytePtr), dataLen);
+            env->ReleaseByteArrayElements(value, bytePtr, 0);
```

Comment:
I believe so; I modeled my use of that off of what you guys had for the `tbl_nativeDoByteArray` method in [`tablebase_tpl.hpp`](https://github.com/realm/realm-java/blob/master/realm/realm-jni/src/tablebase_tpl.hpp), however, I admit that it might not be okay to do?
That being said, I would expect that the unit tests would fail hard if it was releasing it too early.

As far as JNI_ABORT, I can do that for sure.